packages/git/src/node/dugite-git-watcher.ts

```diff
@@ -28,19 +29,31 @@ export class DugiteGitWatcherServer implements GitWatcherServer {
             watcher.dispose();
         }
         this.watchers.clear();
+        this.subscriptions.clear();
     }
 
     async watchGitChanges(repository: Repository): Promise<number> {
         const watcher = this.manager.getWatcher(repository);
-        const disposable = new DisposableCollection();
-        disposable.push(watcher.onStatusChanged(e => {
-            if (this.client) {
-                this.client.onGitChanged(e);
-            }
-        }));
-        disposable.push(watcher.watch());
+
+        const repositoryUri = repository.localUri;
+        let subscriptions = this.subscriptions.get(repositoryUri);
+        if (subscriptions === undefined) {
+            const unsubscribe = watcher.onStatusChanged(e => {
+                if (this.client) {
+                    this.client.onGitChanged(e);
+                }
+            });
+            subscriptions = new DisposableCollection();
+            subscriptions.onDispose(() => {
+                unsubscribe.dispose();
+                this.subscriptions.delete(repositoryUri);
+            });
+            this.subscriptions.set(repositoryUri, subscriptions);
+        }
+
+        const disposable = watcher.watch();
         const watcherId = this.watcherSequence++;
-        this.watchers.set(watcherId, disposable);
+        this.watchers.set(watcherId, subscriptions.push(disposable));
```

Comment:
sorry, my bad, the code before was alright:
```js
const disposable = watcher.watch();
subscriptions.push(disposable); 
this.watchers.set(watcherId, disposable);
```

Reply:
Isn't that the same as the current code ?

```typescript
 const disposable = watcher.watch();
 const watcherId = this.watcherSequence++;    
  this.watchers.set(watcherId, subscriptions.push(disposable));
```

Reply:
No, the trick is, (@akosyakov) can correct me, the latter one will remove itself from the disposable collection too.

Reply:
haaa right I see ok my bad..

Reply:
Just to recap this:
```
 subscriptions.push(disposable)
```
Returns a Disposable that will dispose of the disposable and clean it from the disposable collection.

So if you use:
`this.watchers.set(watcherId, subscriptions.push(disposable));`

You don't need to call ` for (const disposable of this.subscriptions.values()) { disposable.dispose() } `

But if you use:

```
subscriptions.push(disposable); 
this.watchers.set(watcherId, disposable);
```

It doesn't do that, but iterating over the this.watchers and calling dispose then clearing the subscription array is fine.

Reply:
@kittaakos please use the snippet from https://github.com/theia-ide/theia/pull/1164#discussion_r164732624, the current code won't stop watching

Reply:
I think this is just partially correct. My understanding is that you do not have to remove the items explicitly from the disposable collection, but you have to dispose the collection itself.

Reply:
It gets disposed when all disposable are removed.

Reply:
`DisposableCollection.push` returns a disposable to remove a passed disposable. It does not dispose the original disposable, but it gets disposed together with an original disposable.

Reply:
Given that the Diposable dispose method gets overloaded...

Reply:
In the second case, a stored disposable only removes an added disposable from a collection. It does not dispose an original disposable.

Reply:
Haaaa OK yes you're right! thx